command/agent/exec/exec_test.go

```diff
@@ -0,0 +1,288 @@
+				Contents:                 pointerutil.StringPtr(`{{ with secret "kv/my-app/creds" }}{{ .Data.data.user }}{{ end }}`),
+				MapToEnvironmentVariable: pointerutil.StringPtr("MY_USER"),
+			}},
+			testAppArgs:          []string{"--stop-after", "1s", "--exit-code", "5"},
+			testAppStopSignal:    syscall.SIGTERM,
+			testAppPort:          34003,
+			expectedTestDuration: 15 * time.Second,
+			expectedError:        &ProcessExitError{5},
+		},
+
+		"send_sigterm_expect_test_app_exit": {
+			envTemplates: []*ctconfig.TemplateConfig{{
+				Contents:                 pointerutil.StringPtr(`{{ with secret "kv/my-app/creds" }}{{ .Data.data.user }}{{ end }}`),
+				MapToEnvironmentVariable: pointerutil.StringPtr("MY_USER"),
+			}},
+			testAppArgs:                   []string{"--stop-after", "30s", "--sleep-after-stop-signal", "1s"},
+			testAppStopSignal:             syscall.SIGTERM,
+			testAppPort:                   34004,
+			simulatedShutdownWaitDuration: 3 * time.Second,
+			expectedTestDuration:          15 * time.Second,
+		},
+
+		"send_sigusr1_expect_test_app_exit": {
+			envTemplates: []*ctconfig.TemplateConfig{{
+				Contents:                 pointerutil.StringPtr(`{{ with secret "kv/my-app/creds" }}{{ .Data.data.user }}{{ end }}`),
+				MapToEnvironmentVariable: pointerutil.StringPtr("MY_USER"),
+			}},
+			testAppArgs:                   []string{"--stop-after", "30s", "--sleep-after-stop-signal", "1s", "--use-sigusr1"},
+			testAppStopSignal:             syscall.SIGUSR1,
+			testAppPort:                   34005,
+			simulatedShutdownWaitDuration: 3 * time.Second,
+			expectedTestDuration:          15 * time.Second,
+		},
+
+		"test_app_ignores_stop_signal": {
+			envTemplates: []*ctconfig.TemplateConfig{{
+				Contents:                 pointerutil.StringPtr(`{{ with secret "kv/my-app/creds" }}{{ .Data.data.user }}{{ end }}`),
+				MapToEnvironmentVariable: pointerutil.StringPtr("MY_USER"),
+			}},
+			testAppArgs:                   []string{"--stop-after", "60s", "--sleep-after-stop-signal", "60s"},
+			testAppStopSignal:             syscall.SIGTERM,
+			testAppPort:                   34006,
+			simulatedShutdownWaitDuration: 32 * time.Second, // the test app should be stopped immediately after 30s
+			expectedTestDuration:          45 * time.Second,
+		},
+	}
+
+	goBinary, err := exec.LookPath("go")
+	if err != nil {
+		t.Fatalf("could not find go binary on path: %s", err)
+	}
+
+	// we must build a test binary since 'go run' does not propagate signals correctly
+	testAppBinary := filepath.Join(os.TempDir(), "test-app")
+
+	if err := exec.Command(goBinary, "build", "-o", testAppBinary, "./test-app").Run(); err != nil {
+		t.Fatalf("could not build the test application: %s", err)
+	}
+	defer func() {
+		if err := os.Remove(testAppBinary); err != nil {
+			t.Fatalf("could not remove %q test application: %s", testAppBinary, err)
+		}
+	}()
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
```

Comment:
Can we use `t.Parallel()` on the next line here (and at the start of the function)? I can't see anything that might conflict, but I could be missing something

Reply:
I've tried t.Parallel but it's failing the tests -- possibly something related to the data race.

Reply:
Unfortunately, the tests currently can't be run in parallel since there is a shared setup (a common test app that is built ahead of time). This would probably work better in a test suite/fixture, but these don't seem to be used in Vault.